### DISCUSSION TITLE
How to load dynamic Parquet/JSON file list on DuckDB?

### DISCUSSION THREAD

**Discussion by alexhg128** [duckdb/duckdb#8144] [Q&A]
Hey community,

Let's say I got a table that contains a list of JSON files I want to filter and get the JSON files of an specific date (and other factors) then load them using READ_JSON()

The problem I'm facing is that data scanning functions expect constants and not the table in-out I'll get after filtering the JSON filenames. Is there any way around this without doing it through Python? Are data scanning functions expected to accept table in-out in the future?

Thanks!

**Comment by koenvo** [comment]
Curious to your solution.

**Comment by natebenes** [comment]
The `read_json` and `read_parquet` functions expect a list as their first argument.

Suppose you have a query that generates a bunch of URLs:

```
SELECT
  strftime(range, 'https://example.com/files/%Y%m.parquet')
FROM range(date '2022-01-01', date '2023-01-01', interval '1' month);
```

With a mix of lambdas and subqueries, you can reform your result into a list:

```
SELECT
  COUNT(*)
FROM read_parquet(
  list_transform(
    range(date '2022-01-01', date '2023-01-01', interval '1' month),
    x -> strftime(x, 'https://example.com/files/%Y%m.parquet')
  )
);
```

Once you get it working, wrap it in a view. DuckDB 0.8.1 will dynamically build the list each time you `SELECT` from it:

```
CREATE VIEW files2022 AS
  SELECT
    COUNT(*)
  FROM read_parquet(
    list_transform(
      range(date '2022-01-01', date '2023-01-01', interval '1' month),
      x -> strftime(x, 'https://example.com/files/%Y%m.parquet')
    )
  );
```

**Comment by wankdanker** [comment]
**Dynamic file reading from table data using `query()` + `getvariable()`**

Building on @natebenes's `list_transform` approach - that works great when your paths follow a pattern you can generate with `range()` or lambdas. But if your file list comes from an actual table (or CTE, or external API), you hit the wall: `read_json`/`read_parquet` don't support lateral joins or subqueries as arguments.

Here's a workaround using DuckDB's `SET VARIABLE` + `query()` as a table function:

```sql
-- Step 1: Build the dynamic SQL string from your table data
CREATE OR REPLACE TABLE _q AS 
  SELECT 'SELECT *, filename FROM read_json_auto(['
    || string_agg('''' || replace(url, '''', '''''') || '''', ',')
    || '], filename=true, union_by_name=true, sample_size=-1)' AS sql
  FROM (
    SELECT 'https://example.com/api/items/' || id || '.json' AS url
    FROM my_source_table
    WHERE some_filter = true
  );

-- Step 2: Store it as a variable
SET VARIABLE dynamic_sql = (SELECT sql FROM _q);

-- Step 3: Execute via query() as a table function
CREATE OR REPLACE TABLE results AS 
  SELECT * FROM query(getvariable('dynamic_sql'));

-- Now use it like any table
SELECT * FROM results;
```

**Why this works:** `query()` accepts a string and returns a result set you can select from, and `getvariable()` retrieves the variable at parse time so DuckDB sees a literal string. Together they bridge the gap between dynamic table data and the literal-only constraint on scanning functions.

**Tips for reading many files with varying schemas:**
- `union_by_name=true` - merges differing column sets across files, filling missing columns with NULL
- `sample_size=-1` - scans all records for schema inference instead of sampling
- `filename=true` - adds a `filename` column so you can trace each row back to its source
- Use `TRY(struct.field)` or `json_extract(col, '$.field')` to safely access fields that may not exist in every file
----

Note: I did have Claude write this up. It does work and I wanted to share the approach with the community.

**Reply by poikilotherm** [reply]
Thanks @wankdanker this was a great inspiration! For cases where the generated pattern is less dynamic, but also not really predictable, here is my example using `SET VARIABLE` and `getvariable()`, no need for a `query()` here.

```sql
SET VARIABLE export_urls = (
    SELECT list(
        getvariable('base_url')
        || '/api/datasets/export?exporter=dataverse_json&persistentId='
        || pid
    )
    FROM dataset_pids
);

CREATE OR REPLACE TABLE dataset_exports AS
SELECT *
FROM read_json_auto(
    getvariable('export_urls'),
    union_by_name        = true,
    ignore_errors        = true,   -- one bad URL won't abort the whole run
    filename             = true,   -- keeps the source URL per row
    maximum_object_size  = 33554432
);
```
